From 1afd6a68da181e3ea73fb1a8b9163b6e4d0f57b1 Mon Sep 17 00:00:00 2001
From: Omar Polo <op@omarpolo.com>
Date: Mon, 25 Mar 2024 08:55:04 +0000
Subject: [PATCH] configure: add a check to detect Libre/OpenSSL mixings

Now that we're using also some bits from libcrypto, it's important
to not mix different libcrypto/ssl implementation in the same
program.

This check is taken from OpenSMTPD.

May have helped with https://github.com/macports/macports-ports/pull/23007

--- configure.ac
+++ configure.ac
@@ -161,6 +161,22 @@ AC_SEARCH_LIBS([SSL_CTX_new], [ssl], [:], [
 	AC_MSG_ERROR([can't find libssl])
 ])
 
+# Sanity check for Libre/OpenSSL headers vs library mismatch
+AC_MSG_CHECKING([whether libcrypto headers match the library])
+AC_RUN_IFELSE([AC_LANG_PROGRAM([[
+#include <openssl/opensslv.h>
+#include <openssl/crypto.h>
+]], [[
+	return (OpenSSL_version_num() != OPENSSL_VERSION_NUMBER);
+]])], [
+	AC_MSG_RESULT([yes])
+], [
+	AC_MSG_RESULT([no])
+	AC_MSG_ERROR([libcrypto headers do not match the library.])
+], [
+	AC_MSG_WARN([skip (due to cross-compiling)])
+])
+
 AC_CHECK_HEADERS([linux/landlock.h])
 
 dnl after all the function checks, add optional support for -Werror
